graphqlapi/src/main/java/io/stargate/graphql/web/resources/GraphqlCache.java

```diff
@@ -232,70 +220,99 @@ public void onKeyspaceChanged(
   }
 
   /**
-   * Holds the GraphQL schema for a particular keyspace (either CQL-first or GraphQL-first,
-   * depending on whether there is a SchemaSource for this keyspace).
+   * An entry that holds the GraphQL schema cached for a particular keyspace (either CQL-first or
+   * GraphQL-first, depending on whether a custom schema was deployed).
    */
-  class DmlGraphqlHolder {
+  interface GraphqlHolder {
+    GraphQL getGraphql();
 
-    // the user-provided GraphQL schema, or null if CQL-first
-    private final SchemaSource source;
-    private final Keyspace keyspace;
-    // This future is used to guarantee that the cached value is built only once, even if two
-    // clients try to initialize it concurrently.
-    private final CompletableFuture<GraphQL> graphqlFuture = new CompletableFuture<>();
+    /**
+     * Whether the schema matches the given source (if it doesn't, it will have to be recomputed).
+     * Note that the source can be {@code null} to indicate that no custom schema was deployed.
+     */
+    boolean matches(@Nullable SchemaSource source);
 
-    DmlGraphqlHolder(SchemaSource source, Keyspace keyspace) {
-      this.source = source;
-      this.keyspace = keyspace;
-    }
+    boolean isCqlFirst();
+  }
 
-    DmlGraphqlHolder(SchemaSource source, GraphQL graphQL) {
+  /**
+   * An entry used when we need to recompute the schema (because we're reacting to the insertion of
+   * a new row in the source table).
+   */
+  class LazyGraphqlHolder implements GraphqlHolder {
+    private final SchemaSource source;
+    private final Supplier<GraphQL> computeGraphql;
+    private volatile GraphQL graphql;
+
+    LazyGraphqlHolder(SchemaSource source, Keyspace keyspace) {
       this.source = source;
-      this.graphqlFuture.complete(graphQL);
-      // when the graphqlFuture is already completed, the keyspace is not-needed
-      this.keyspace = null;
+      this.computeGraphql =
+          (source == null)
+              ? () -> newGraphql(SchemaFactory.newDmlSchema(keyspace))
+              : () -> {
+                ProcessedSchema processedSchema =
+                    new SchemaProcessor(persistence, true).process(source.getContents(), keyspace);
+                // Check that the data model still matches
+                CassandraMigrator.forPersisted()
+                    .compute(processedSchema.getMappingModel(), keyspace);
+                return processedSchema.getGraphql();
+              };
     }
 
-    void init() {
-      try {
-        graphqlFuture.complete(buildGraphql());
-      } catch (Exception e) {
-        graphqlFuture.completeExceptionally(e);
+    @Override
+    public GraphQL getGraphql() {
+      GraphQL result = graphql;
```

Comment:
you can use the suppliers from the guava library that calculate the result only once in order to not care about the concurrency here: https://guava.dev/releases/29.0-jre/api/docs/com/google/common/base/Suppliers.html#memoize-com.google.common.base.Supplier-

Reply:
Great idea, will do.